### PR TITLE
Concurrent traversal

### PR DESCRIPTION
**Pull request by roysc**
Implements concurrent iteration of the state trie via the utilities in https://github.com/vulcanize/go-eth-state-node-iterator/pull/4 and a new command parameter.

### CODE REVIEW

**Comment by roysc** [comment]
Currently this takes a parameter called `divide-depth` _d_ and splits the domain into 16^_d_ intervals, but it can be made more flexible to split into a specific number of jobs.

**Comment by i-norden** [comment]
Haven't had a chance to really review the code but I went ahead and tested running it. 

With divideDepth == 0 

```
Ians-MacBook-Air:eth-pg-ipfs-state-snapshot iannorden$ time ./eth-pg-ipfs-state-snapshot stateSnapshot --config=./environments/example.toml 
INFO[2020-08-24T08:58:12-05:00] ----- Starting vDB -----                     
INFO[2020-08-24T08:58:12-05:00] Using config file: ./environments/example.toml 
INFO[2020-08-24T08:58:12-05:00] Log level set to info                        
INFO[2020-08-24T08:58:12-05:00] Creating snapshot at head                    
INFO[2020-08-24T08:58:12-05:00] Creating snapshot at height 77235            
INFO[2020-08-24T09:02:35-05:00] state snapshot at height -1 is complete       SubCommand=stateSnapshot

real	4m22.710s
user	0m17.982s
sys	0m22.169s
```

With divideDepth == 1 

```
Ians-MacBook-Air:eth-pg-ipfs-state-snapshot iannorden$ time ./eth-pg-ipfs-state-snapshot stateSnapshot --config=./environments/example.toml 
INFO[2020-08-24T09:03:00-05:00] ----- Starting vDB -----                     
INFO[2020-08-24T09:03:00-05:00] Using config file: ./environments/example.toml 
INFO[2020-08-24T09:03:00-05:00] Log level set to info                        
INFO[2020-08-24T09:03:00-05:00] Creating snapshot at head                    
INFO[2020-08-24T09:03:00-05:00] Creating snapshot at height 77235            
INFO[2020-08-24T09:03:27-05:00] state snapshot at height -1 is complete       SubCommand=stateSnapshot

real	0m27.402s
user	0m9.237s
sys	0m8.268s
```

That's a significant speed up! 

Only issue is that when I increase the divideDepth above 1, it begins to throws some errors:

With divideDepth == 2 

```
Ians-MacBook-Air:eth-pg-ipfs-state-snapshot iannorden$ time ./eth-pg-ipfs-state-snapshot stateSnapshot --config=./environments/example.toml 
INFO[2020-08-24T09:04:24-05:00] ----- Starting vDB -----                     
INFO[2020-08-24T09:04:24-05:00] Using config file: ./environments/example.toml 
INFO[2020-08-24T09:04:24-05:00] Log level set to info                        
INFO[2020-08-24T09:04:24-05:00] Creating snapshot at head                    
INFO[2020-08-24T09:04:24-05:00] Creating snapshot at height 77235            
FATA[2020-08-24T09:04:24-05:00] dial tcp: lookup localhost: no such host      SubCommand=stateSnapshot

real	0m0.326s
user	0m0.250s
sys	0m0.151s
```

and sometimes the error is this message instead

```
Ians-MacBook-Air:eth-pg-ipfs-state-snapshot iannorden$ time ./eth-pg-ipfs-state-snapshot stateSnapshot --config=./environments/example.toml 
INFO[2020-08-24T09:05:15-05:00] ----- Starting vDB -----                     
INFO[2020-08-24T09:05:15-05:00] Using config file: ./environments/example.toml 
INFO[2020-08-24T09:05:15-05:00] Log level set to info                        
INFO[2020-08-24T09:05:15-05:00] Creating snapshot at head                    
INFO[2020-08-24T09:05:15-05:00] Creating snapshot at height 77235            
FATA[2020-08-24T09:05:15-05:00] write tcp [::1]:53978->[::1]:5432: write: broken pipe  SubCommand=stateSnapshot

real	0m0.365s
user	0m0.271s
sys	0m0.165s
```

For some reason we lose our ability to connect to Postgres over localhost:5432. I'm not entirely sure what is causing it but I suspect we are exceeding the maximum number of allowed connections (default is ~100).

**Comment by i-norden** [comment]
@arijitAD I think it would be good for you to review this and pick it back up, the speed ups were very significant in our small tests.

**Comment by roysc** [comment]
Updated with some tests and tooling.

**Comment by i-norden** [comment]
Thanks Roy, is this ready for a final review?

**Comment by roysc** [comment]
Yes, I think so. Tests are rudimentary but they can be expanded.
I've tested manually too, but only at small scale. I'll pull down more chain data and check again.

**Comment by roysc** [comment]
Actually, this still needs a storage node test case. I should download/convert it but I don't think it needs to block this if we manually verify it.

**Review by i-norden — APPROVED** [review]
LGTM, we do need to test the storage trie(s) and also code => codeHash mappings. And some benchmarks unit tests to compare relative performance before and after these updates + the up-coming csv/sql file writing changes would be ideal. Did you have a chance to run benchmarks manually, is it similar to the ones shown in my comment above?

File: go.mod [hunk 1]
```diff
@@ -41,3 +43,5 @@ require (
 )
 
 replace github.com/ethereum/go-ethereum v1.9.14 => github.com/vulcanize/go-ethereum v1.10.14-statediff-0.0.29
+
+replace github.com/vulcanize/go-eth-state-node-iterator => ../state-node-iterator
```
Comment (i-norden) [inline go.mod:47]: This should be replaced before merging

Comment (i-norden) [reply go.mod:47]: @roysc if you push up your branch of the state node iterator with the mocks pkg, I can resolve the dep issue, perform some local benchmarks and get this merged.

**Pull request merged**